conversant/prompt_chatbot.py

```diff
@@ -244,79 +257,127 @@ def generate_prompt_update_examples(self, query: str) -> str:
-            self.prompt_size_history[-1] += self.co.tokenize(response).length
-
-    def partial_reply(self, query: str) -> str:
+    def partial_reply(self, query: str) -> Tuple[str, str]:
         """Generates (partial) reply to a query given a chat history.
 
         Args:
             query (str): A query passed to the prompt chatbot.
 
         Yields:
-            str: Dictionary of query and generated LLM response
+
+            Tuple[str, str]: A tuple of the response before the co.generate call,
+                and the response after.
         """
         current_prompt = self.generate_prompt_update_examples(query)
+        self.prompt_history.append(current_prompt)
+
+        response_before_current = ""
         response_so_far = ""
-        should_break = False
-        for i in range(self.partial_reply_max_reruns()):
-            if not should_break:
-                generated_object = self.co.generate(
-                    model=self.client_config["model"],
-                    prompt=current_prompt,
-                    max_tokens=TOKENS_PER_REQUEST,
-                    temperature=self.client_config["temperature"],
-                    frequency_penalty=self.client_config["frequency_penalty"],
-                    presence_penalty=self.client_config["presence_penalty"],
-                    stop_sequences=self.client_config["stop_sequences"],
-                )
-                response = generated_object.generations[0].text
-
-                # Fetches the stop sequence at the end of response if it exists
-                stop_seq = self.get_stop_seq(response)
-                if stop_seq != "" or response == "":
-                    # If there is a stop sequence at the end of response
-                    # remove it and set should_break flag to True
-                    if stop_seq != "":
-                        response = response[: -len(stop_seq)]
-                        should_break = True
-                    # If response is empty, break immediately
-                    else:
-                        break
-                current_prompt += response
-                response_so_far += response
+        num_requests_made = 0
+        max_requests = int(self.client_config["max_tokens"] / TOKENS_PER_REQUEST)
+        reply_complete = False
+
+        # As soon as the function is called (and the generator is created), dispatch
+        # a concurrent call to co.generate
+        future = self._dispatch_concurrent_generate_call(
+            model=self.client_config["model"],
+            prompt=current_prompt,
+            max_tokens=TOKENS_PER_REQUEST,
+            temperature=self.client_config["temperature"],
+            frequency_penalty=self.client_config["frequency_penalty"],
+            presence_penalty=self.client_config["presence_penalty"],
+            stop_sequences=self.client_config["stop_sequences"],
+        )
+
+        while num_requests_made < max_requests and not reply_complete:
```

Comment:
Factoring out to a while loop so that there are no additional conditionals that check for `should_break`/`reply_complete`